### PR TITLE
feat(LOC-2232): maintain UI state between sites

### PR DESCRIPTION
**Pull request by veryspry**
### Summary

Currently, many UI elements were not being properly maintained between site views. This was primarily due to the fact that IPC listeners were created inside of a React component and would consequently be set up and torn down when switching site views. The result of this was that events emitted by the main thread didn't have any listeners in renderer if a corresponding React component had been unmounted and then the redux store was not getting updated correctly.

Also fixes up a few other bugs and makes some small improvements including:
- fix bug where progress bar wasn't accurately showing the compression progress in all cases
- remember wether the overview or file list is active for each site and automatically show the previous view. (only during a single Local runtime. If Local is restarted, the overview will be automatically shown.)
- adds ability to cancel compression on a per site basis. This was necessary as the previous behavior would not have allowed cancelling if more than one site was running a compression process at the same time.
- fixes a bug where the displayed compression savings on the file list was wrong if the compression process had been cancelled. 

I kept the commits on this one nice and focused, so feel free to dig through on a per commit basis if you want to see isolated changes.

### CODE REVIEW

File: src/renderer/setupListeners.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			},
+		},
+	];
+
+	listeners.forEach(({ channel, cb }) => {
+		if (!ipcRenderer.listenerCount(channel)) {
+			ipcRenderer.on(channel, cb);
+		}
+	});
+}
```
Comment (tihal) [inline src/renderer/setupListeners.ts:51]: Very nice to have these split out into their own file!

File: src/renderer/fileListView/fileListHeader.tsx [hunk 2]
```diff
@@ -66,12 +64,6 @@ export const FileListHeader = (props: IFileListHeaderProps) => {
     } else if (siteImageData.optimizationStatus === OptimizerStatus.RUNNING) {
         return (
             <div className='fileView_Header'>
-				<NavigationPrompt
-					when={true}
-				>
-					{({ onConfirm, onCancel }) => invokeModal(WarningModal as React.FC, onConfirm, onCancel)}
-				</NavigationPrompt>
-
```
Comment (tihal) [inline src/renderer/fileListView/fileListHeader.tsx:74]: So glad we can ditch this.

File: src/renderer/index.tsx [hunk 3]
```diff
@@ -41,74 +37,9 @@ const ImageOptimizer = (props: IProps) => {
...
-				ipcRenderer.removeAllListeners(IPC_EVENTS.COMPRESS_ALL_IMAGES_COMPLETE);
-				ipcRenderer.removeAllListeners(IPC_EVENTS.SCAN_IMAGES_COMPLETE);
-				ipcRenderer.removeAllListeners(IPC_EVENTS.SCAN_IMAGES_FAILURE);
-			}
-		},
-		[siteImageData],
-	);
+	const setOverviewSelected = (overviewSelected: boolean) => {
+		store.dispatch(actions.overviewSelected({ siteID, overviewSelected }));
+	}
```
Comment (tihal) [inline src/renderer/index.tsx:42]: As far as I can tell, this is the state object used to persist whether a scan is running or not, correct?

Comment (veryspry) [reply src/renderer/index.tsx:42]: Correct!

Comment (veryspry) [reply src/renderer/index.tsx:42]: Added a field to the `SiteImageData` type: https://github.com/getflywheel/local-addon-image-optimizer/pull/48/files#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410aR29

File: src/renderer/fileListView/fileListHeader.tsx (hunk 2 above)
Comment (veryspry) [reply src/renderer/fileListView/fileListHeader.tsx:74]: Same

**Review by tihal — APPROVED** [review]
Reviewed on a call w/ Crum. Approval from both of us ✅

**Pull request merged**